ignite/contrib/handlers/time_profilers.py

```diff
@@ -437,3 +438,299 @@ def odict_to_str(d: Mapping):
         )
         print(output_message)
         return output_message
+
+
+class HandlersTimeProfiler:
+    """
+    HandlersTimeProfiler can be used to profile the handlers,
+    data loading and data processing times.
+
+    Examples:
+
+    .. code-block:: python
+
+        from ignite.contrib.handlers import HandlersTimeProfiler
+
+        trainer = Engine(train_updater)
+
+        # Create an object of the profiler and attach an engine to it
+        profiler = HandlersTimeProfiler()
+        profiler.attach(trainer)
+
+        @trainer.on(Events.EPOCH_COMPLETED)
+        def log_intermediate_results():
+            profiler.print_results(profiler.get_results())
+
+        trainer.run(dataloader, max_epochs=3)
+
+        profiler.write_results('path_to_dir/time_profiling.csv')
+
+    """
+
+    def __init__(self):
+        self._dataflow_timer = Timer()
+        self._processing_timer = Timer()
+        self._event_handlers_timer = Timer()
+
+        self.dataflow_times = None
+        self.processing_times = None
+        self.event_handlers_times = None
+
+    @staticmethod
+    def _get_callable_name(handler: Callable):
+        # get name of the callable handler
+        return handler.__qualname__ if hasattr(handler, "__qualname__") else handler.__class__.__name__
+
+    def _timeit_handler_wrapper(self, handler: Tuple[Callable, List[Any], Mapping[str, Any]]):
+        def _timeit_handler(event: Events, *args: Any, **kwargs: Any):
+            # TODO: add threshold filtering for handlers with event filter
+            # if (
+            #     isinstance(event_name, CallableEventWithFilter)
+            #     and event_name.filter != CallableEventWithFilter.default_event_filter
+            # ):
+            self._event_handlers_timer.reset()
+            func, args, kwargs = handler
+            func(*args, **kwargs)
+            t = self._event_handlers_timer.value()
+            hname = self._get_callable_name(func)
+            self.event_handlers_times[event][hname].append(t)
+
+        # required to revert back to original handler after profiling
+        _timeit_handler._profiler_original = handler
```

Comment:
```suggestion
        setattr(_timeit_handler, "_profiler_original", handler)
```
Anticipate later work for mypy.